### PR TITLE
Added required attribute to ScenarioState.

### LINKED ISSUE #255: Introduce mandatory Scenario State
It would be useful to have an attribute `mandatory = true` in the `@ScenarioState` annotation. If this is set, JGiven will fail if no such `@ScenarioState` would have been provided.

Example:
```
class MyStage {
    @ScenarioState(mandatory = true)
    String mandatoryState;
}
```

### PR DESCRIPTION
**Pull request by Airblader**
This commit adds a new boolean attribute "required" to both
ScenarioState and ExpectedScenarioState. If set to true on a
field within a stage, corresponding tests will fail automatically
if the state hasn't been provided.

fixes #255

### CODE REVIEW

**Comment by Airblader** [comment]
Clearly not ready to merge yet. @janschaefer I assume you prefer a `JGivenWrongUsageException`, right? Do you think the basic approach is OK?

And, lastly, under what version should I make the changelog entry?

**Comment by janschaefer** [comment]
I am currently wondering whether one could also have a required provided. So it would fail if a stage does not provide a certain value when it's finished. What do you think?

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java [hunk 1]
```diff
@@ -90,6 +90,8 @@ public void updateValues( Object object ) {
                 if( value != null ) {
                     field.set( object, value );
                     log.debug( "Setting field {} to value {}", field, value );
+                } else if( isRequired( field ) ) {
+                    throw new RuntimeException( "TODO Better exception?" );
```
Comment (janschaefer) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: Yeah, this should be a different exception. The question is which kind of exception. `JGivenWrongUsageException` is not really fitting, I think, because JGiven is not really used in a wrong way in this case. JGivenPreconditionException/JGivenAssumptionException/JGivenUserException maybe?

**Comment by Airblader** [comment]
It definitely makes sense, in particular because a dependent stage wouldn't have to look at the implementation to see whether the state will always be provided or under which circumstances it will be provided. 

I think we should give this feature more thought in that case. Currently these annotations behave the same and this would make them actually do different things (instead of one of them just not having all the features of the others). 
We either need to accept this "mental breakage" (and should remove the documentation claiming they all behave the same) or make it explicit by naming the attributes differently. Otherwise I'm afraid it becomes a little unintuitive.

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java [hunk 2]
```diff
@@ -90,6 +90,8 @@ public void updateValues( Object object ) {
                 if( value != null ) {
                     field.set( object, value );
                     log.debug( "Setting field {} to value {}", field, value );
+                } else if( isRequired( field ) ) {
```
Comment (janschaefer) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: I wonder whether the code can be optimized so that it does not have to iterate twice over the annotations of the field. Because the getResolution method also does the iteration.

File: jgiven-junit/src/test/java/com/tngtech/jgiven/junit/ScenarioExecutionTest.java [hunk 3]
```diff
@@ -352,4 +352,17 @@ public void abstract_steps_should_appear_in_the_report_model() throws Throwable
 
     }
 
+    static class StageWithMissingState {
+        @ScenarioState( required = true )
```
Comment (janschaefer) [inline jgiven-junit/src/test/java/com/tngtech/jgiven/junit/ScenarioExecutionTest.java]: you could also add a test for `@ExpectedScenarioState`. Maybe also interesting would be a test case where the stage is not the first one

Comment (Airblader) [inline jgiven-junit/src/test/java/com/tngtech/jgiven/junit/ScenarioExecutionTest.java]: Agree, more tests would be good here.

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java (hunk 1 above)
Comment (Airblader) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: Sure. I'll come up with something. :-)

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java (hunk 2 above)
Comment (Airblader) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: I think the answer to »can it« is a clear yes. I'm not sure about whether it's worth it, though. Fields generally don't have dozens, hundreds or thousands of annotations, so my first guess is that we're optimizing a non-issue here. Do you have specific concerns about performance for this?

**Comment by janschaefer** [comment]
Yes, I deleted the other comment, because I actually wanted to add that to the review :-). Why not calling the attribute `mandatory`? Then it is universal and can be used in `BeforeScenarioState` and `AfterScenarioState`?

**Comment by janschaefer** [comment]
Did you post your comments twice?

File: jgiven-junit/src/test/java/com/tngtech/jgiven/junit/ScenarioExecutionTest.java (hunk 3 above)
Comment (janschaefer) [inline jgiven-junit/src/test/java/com/tngtech/jgiven/junit/ScenarioExecutionTest.java]: In particular, there should be a test with multiple steps where the second step sets the value.

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java (hunk 2 above)
Comment (janschaefer) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: Well, in general I agree, however, JGiven is already doing a lot of reflection stuff, so I like to keep it at a minimum if possible. But I guess it will not be measurable.

**Comment by Airblader** [comment]
> Did you post your comments twice?

No, but I saw it too. But as fast as they appeared they also went away again. Replication issues @ Github, perhaps. :-)

> Why not calling the attribute mandatory?

My feeling was that `required` is more of an idiom here. I couldn't think of any Java library that uses the »mandatory« terminology. Granted, this is highly subjective. :-) Ultimately they're synonyms, anyway.

> Then it is universal and can be used in BeforeScenarioState and AfterScenarioState?

(Assuming you meant to say ExpectedScenarioState / ProvidedScenarioState)

I don't think using `mandatory` solves my concerns. We'd have three annotations which look similar, claim to behave exactly the same but have slightly different semantics for the same interface. This could easily confuse users and, e.g., lead them to believe that ProvidedScenarioState has a meaningless attribute.

By the way, is there a common practice (»Java-wide« or just within JGiven) when to use attributes versus dedicated annotations? I could also imagine having `@Required` (and something like `@Guaranteed` for the other case, perhaps).

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java (hunk 2 above)
Comment (Airblader) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: The obvious approach here would be to encapsulate annotations so we iterate over them once and »cache« the information we collect. I don't mind adding this if you prefer it.

**Comment by janschaefer** [comment]
I agree, if you have a boolean attribute that you then as well can have another annotation. I am not aware of any common practice here. I think that is a matter of taste. However, in this case, it would make the implementation easier, because we would not have to add the attribute to any of the annotations, so the annotations would all still could behave the same. So we would have something like:

    @ScenarioState @Required
    String name;

    @ScenarioState @Guaranteed
    String message;

The semantics could then be defined for the two new annotations, independently of the three `ScenarioState` annotations/

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java (hunk 2 above)
Comment (janschaefer) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: Yes that makes sense. Please do so :-)

**Comment by Airblader** [comment]
Fun fact, I tried to do some research on mandatory versus required.

* On Github, 355+ million results for required, but only 15+ million for mandatory.
* On Debian Code Search, 1.5+ million for required and 100k for mandatory.
* According to some linguists, mandatory is primarily used in legal work. Unfortunately no similar statement about required.

---

Yeah, I think dedicated annotations make more sense here for the reasons you mentioned. I'll work all of the feedback in. I'll consider whether I'll add `@Guaranteed` right away or do a separate PR for it. Need to set up Eclipse again first. So far I did this in vim… :-)

**Comment by janschaefer** [comment]
Interesting and mandetory is also harder to spell ;-).
Regarding the IDE: I suggest IntelliJ in this case, because the project generation with ./gradlew idea works out of the box.

**Comment by janschaefer** [comment]
I agree to implement @Guaranteed in another PR.

**Comment by Airblader** [comment]
@janschaefer Actually, can you be more specific on the other test you want to see? Or just find me and we'll talk about it. J

Things left to do:
-[ ] Adapt CHANGELOG (for which version?)
-[ ] Add the missing test case(s)

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java [hunk 4]
```diff
@@ -42,57 +43,65 @@ public void validateFields( Object object ) {
 
         Map<Object, Field> resolvedFields = Maps.newHashMap();
 
-        for( Field field : getScenarioFields( object ) ) {
-            field.setAccessible( true );
-            Resolution resolution = getResolution( field );
+        for( ScenarioStateField field : getScenarioFields( object ) ) {
```
Comment (Airblader) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: Now we go through annotations each time `getScenarioFields` is called, which still happens multiple times even within one method. I think refactoring this would deserve its own PR, though, in particular since this has been existng behavior.

File: jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java [hunk 5]
```diff
@@ -42,57 +43,65 @@ public void validateFields( Object object ) {
...
-            try {
-                Object value = getValue( field );
-                if( value != null ) {
-                    field.set( object, value );
-                    log.debug( "Setting field {} to value {}", field, value );
+        for( ScenarioStateField field : getScenarioFields( object ) ) {
+            Object value = getValue( field );
+
+            if( value != null ) {
+                try {
```
Comment (Airblader) [inline jgiven-core/src/main/java/com/tngtech/jgiven/impl/inject/ValueInjector.java]: Moved these try/catch blocks around so the IllegalAccess block only encapsulates the code that actually can throw this.

**Comment by janschaefer** [comment]
Version would be 0.14.0

**Comment by Airblader** [comment]
@janschaefer We didn't get to continue our conversation today about attribute versus dedicated annotation. :-)

**Comment by janschaefer** [comment]
ok, I thought we agreed on an attribute :-)

**Comment by janschaefer** [comment]
@Airblader: I reverted the use of the ShadowJar plugin. This caused your PR to have merge conflicts now.

**Comment by Airblader** [comment]
@janschaefer Already resolving. I can deal with a few conflicts. :-)

**Comment by janschaefer** [comment]
Wow, you are faster than your Shadow :-D. I already resolved them :-)

**Comment by Airblader** [comment]
@janschaefer Oh, so did I and force-pushed my changes because I did a rebase instead of a merge. :-)

**Comment by janschaefer** [comment]
Ok, that is better :-)

**Comment by Airblader** [comment]
Only that I missed something. :-( I'll update again in a second.

**Comment by Airblader** [comment]
Should be all good now. Just gotta make Eclipse work again… :-)

**Comment by janschaefer** [comment]
One last thing: could you document this feature in the user guide?
